group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,413 @@
+
+        unassignedPartitions.forEach(partition -> {
+            boolean assigned = false;
+
+            if (rackInfo.useRackStrategy && currentPartitionOwners.containsKey(partition)) {
+                String prevOwner = currentPartitionOwners.get(partition);
+                if (unfilledMembers.containsKey(prevOwner)) {
+                    assignPartitionToMember(prevOwner, partition);
+                    assigned = true;
+                    if (!unfilledMembers.containsKey(prevOwner)) {
+                        roundRobinMembers.remove(prevOwner);
+                    }
+                }
+            }
+
+            // Only re-add the member to the end of the queue if it's still available for assignment.
+            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                if (unfilledMembers.get(memberId) > 0) {
+                    assignPartitionToMember(memberId, partition);
+                    assigned = true;
+                }
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+        });
+    }
+
+    /**
+     * Assigns the specified partition to the given member and updates the unfilled members map.
+     *
+     * <p>
+     * If the member has met their allocation quota, the member is removed from the unfilled members map.
+     * Otherwise, the count of remaining partitions that can be assigned to the member is updated.
+     * </p>
+     *
+     * @param memberId              The Id of the member to which the partition will be assigned.
+     * @param topicIdPartition      The topicIdPartition to be assigned.
+     */
+    private void assignPartitionToMember(String memberId, TopicIdPartition topicIdPartition) {
+        addPartitionToAssignment(
+            topicIdPartition.partition(),
+            topicIdPartition.topicId(),
+            memberId,
+            targetAssignment
+        );
+
+        int remainingPartitionCount = unfilledMembers.get(memberId) - 1;
+        if (remainingPartitionCount == 0) {
+            unfilledMembers.remove(memberId);
+        } else {
+            unfilledMembers.put(memberId, remainingPartitionCount);
+        }
+    }
+
+    /**
+     * Determines which members can still be assigned partitions to meet the full quota.
+     *
+     * The map contains:
+     * <ol>
+     *     <li> Members that have met the minimum quota but will receive an extra partition. </li>
+     *     <li> Members that have not yet met the minimum quota. If there are still members that could receive
+     *          an extra partition, the remaining value is updated to include this. </li>
+     * </ol>
+     *
+     * @return A map of member Ids and the required partitions to meet the quota.
+     */
+    private Map<String, Integer> computeUnfilledMembers() {
+        Map<String, Integer> unfilledMembers = new HashMap<>();
```

Comment:
I was wondering whether we could avoid the copy and just update `potentiallyUnfilledMembers`. Have you thought about this?

Reply:
Yeah you're right, it's an unnecessary extra map, I made it a local map, just wanna still keep it separate cause potentially unfilled members also has ones that have met the quota, but unfilled only has those that still need to be assigned partitions, might be hard to understand what types of members we have at each point. Lmk if what I did makes sense or if we can optimize it more

Reply:
yeah that's true I didn't think of how we're assigning the extra partitions, used the same logic as before, let me look into it. Thanks for catching this!

Reply:
I think one option would be to iterate through potentially unfilled members instead during the round robin process and if the member is a potential rack match plus can get an extra partition then we can assign it. We basically dynamically decide who gets the extra, instead of deciding before hand

Reply:
Yeah, I was thinking about more or less the same. Is it an small change?

Reply:
Wasn't sadly but its done now